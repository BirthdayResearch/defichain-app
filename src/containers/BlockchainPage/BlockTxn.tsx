import React, { Component } from 'react';
import {
  Button,
  Card
} from 'reactstrap';
import {
  MdContentCopy,
  MdInfo,
  MdArrowForward,
} from 'react-icons/md';
import styles from './BlockTxn.module.scss';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import classnames from 'classnames';

class BlockTxn extends Component<any,any> {
  state = {
    copied: false
  }

  handleCopy = () => {
    this.setState({
      copied: true
    });
    setTimeout(() => {
      this.setState({
        copied: false
      });
    }, 600);
  }

  render() {
    const txn = this.props.txn;
    const froms = txn.froms;
    let fromsRender;

    if (froms.length) {
      fromsRender = (
        froms.map((from) => (
          <div className={styles.from}>
            <span className={styles.address}>
              {from.address ? from.address : "Unparsed address"}
            </span>
            <span className={styles.amount}>
              {from.amount} DFI
            </span>
          </div>
        ))
      );
    } else {
      fromsRender = (
        <div className={styles.fromsEmpty}>
          No Inputs (Newly Generated Coins)
        </div>
      )
    }

    return (
      <Card className={`${styles.txn} styles.txn`}>
        <div className={styles.header}>
          <span className={styles.hash}>
            <div className={classnames({ 'd-block': this.state.copied }, styles.copiedIndicator)}>Copied!</div>
            <span className={styles.hashSpan}>{txn.hash}</span>
            <CopyToClipboard text={txn.hash}>
              <Button
                color="link" size="sm"
                className="padless ml-2"
                onClick={this.handleCopy}
              >
                <MdContentCopy />
              </Button>
            </CopyToClipboard>
          </span>
          <span className={styles.time}>
            <span className={styles.timeSpan}>{txn.time}</span>
            <Button
              color="link" size="sm"
              className="padless ml-2"
            >
              <MdInfo />
            </Button>
          </span>
        </div>
        <div className={styles.addresses}>
          <div className={styles.froms}>
            {fromsRender}
          </div>
          <div className={styles.arrow}>
            <MdArrowForward />
          </div>
          <div className={styles.tos}>
            {txn.tos.map((to) => (
              <div className={styles.to}>
                <span className={styles.address}>
                  {to.address ? to.address : "Unparsed address"}
                </span>
                <span className={styles.amount}>
                  {to.amount} DFI
                </span>
              </div>
            ))}
          </div>
        </div>
      </Card>
    );
  }
}

export default BlockTxn;