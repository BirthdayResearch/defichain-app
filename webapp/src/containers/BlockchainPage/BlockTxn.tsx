import React, { Component } from "react";
import { Button, Card } from "reactstrap";
import { MdContentCopy, MdInfo, MdArrowForward } from "react-icons/md";
import styles from "./BlockTxn.module.scss";
import CopyToClipboard from "react-copy-to-clipboard";
import classnames from "classnames";
import { I18n } from "react-redux-i18n";

interface BlockTxnProps {
  txn: {
    hash: string;
    time: string;
    froms: Array<{
      address: string;
      amount: number | string;
    }>;
    tos: Array<{
      address: String;
      amount: number | string;
    }>;
  };
}

interface BlockTxnState {
  copied: boolean;
}

class BlockTxn extends Component<BlockTxnProps, BlockTxnState> {
  state = {
    copied: false,
  };

  handleCopy = () => {
    this.setState({
      copied: true,
    });
    setTimeout(() => {
      this.setState({
        copied: false,
      });
    }, 600);
  };

  render() {
    const txn = this.props.txn;
    const froms = txn.froms;
    let fromsRender;

    if (froms.length) {
      fromsRender = froms.map((from) => (
        <div className={styles.from}>
          <span className={styles.address}>
            {from.address
              ? from.address
              : I18n.t("containers.blockChainPage.blockTxn.unparsedAddress")}
            }
          </span>
          <span className={styles.amount}>
            {from.amount} {I18n.t("containers.blockChainPage.blockTxn.dFI")}
          </span>
        </div>
      ));
    } else {
      fromsRender = (
        <div className={styles.fromsEmpty}>
          {I18n.t("containers.blockChainPage.blockTxn.noInputs")}
        </div>
      );
    }

    return (
      <Card className={`${styles.txn} styles.txn`}>
        <div className={styles.header}>
          <span className={styles.hash}>
            <div
              className={classnames(
                { "d-block": this.state.copied },
                styles.copiedIndicator
              )}
            >
              {I18n.t("containers.blockChainPage.blockTxn.copied")}
            </div>
            <span className={styles.hashSpan}>{txn.hash}</span>
            <CopyToClipboard text={txn.hash}>
              <Button
                color="link"
                size="sm"
                className="padless ml-2"
                onClick={this.handleCopy}
              >
                <MdContentCopy />
              </Button>
            </CopyToClipboard>
          </span>
          <span className={styles.time}>
            <span className={styles.timeSpan}>{txn.time}</span>
            <Button color="link" size="sm" className="padless ml-2">
              <MdInfo />
            </Button>
          </span>
        </div>
        <div className={styles.addresses}>
          <div className={styles.froms}>{fromsRender}</div>
          <div className={styles.arrow}>
            <MdArrowForward />
          </div>
          <div className={styles.tos}>
            {txn.tos.map((to) => (
              <div className={styles.to}>
                <span className={styles.address}>
                  {to.address
                    ? to.address
                    : I18n.t(
                        "containers.blockChainPage.blockTxn.unparsedAddress"
                      )}
                </span>
                <span className={styles.amount}>
                  {to.amount} {I18n.t("containers.blockChainPage.blockTxn.dFI")}
                </span>
              </div>
            ))}
          </div>
        </div>
      </Card>
    );
  }
}

export default BlockTxn;
