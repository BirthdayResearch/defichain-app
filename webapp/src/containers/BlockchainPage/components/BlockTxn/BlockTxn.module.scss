// App and Bootstrap variables required
@import '~bootstrap/scss/functions';
@import '~bootstrap/scss/variables';
@import '~bootstrap/scss/mixins';
@import '../../../../scss/variables';
@import '../../../../scss/bootstrap/variables';

.txn {
  margin-bottom: 1.5rem;
  padding-left: $table-cell-padding-x;
}

.header {
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-between;
  align-items: center;
  padding: $table-cell-padding-y $table-cell-padding-x $table-cell-padding-y 0;
  border-bottom: $table-border-width solid $table-border-color;
  font-size: 0.875rem;
  line-height: 1.5rem;
}

.hash {
  position: relative;
  max-width: 60%;
  padding-right: $table-cell-padding-x;
  white-space: nowrap;
  display: flex;
}

.hashSpan,
.timeSpan {
  display: block;
  text-overflow: ellipsis;
  overflow: hidden;
}

.time {
  max-width: 40%;
  white-space: nowrap;
  display: flex;
}

.copiedIndicator {
  position: absolute;
  top: 0;
  left: 0;
  width: calc(100% - 48px);
  height: 100%;
  background-color: $card-bg;
  text-align: center;
  display: none;
}

.addresses {
  display: flex;
  align-items: center;
  font-size: 0.875rem;
  line-height: 1.25rem;
  padding-top: $table-cell-padding-y * 0.5;
  padding-bottom: $table-cell-padding-y;
}

.froms,
.tos {
  flex-grow: 1;
  width: calc(50% - 12px);
}

.froms {
  padding-right: $table-cell-padding-x;
}

.tos {
  padding-left: $table-cell-padding-x;
  padding-right: $table-cell-padding-x;
}

.arrow {
  flex-shrink: 0;
  width: 24px;
  height: 24px;

  svg {
    width: 24px;
    height: 24px;
    vertical-align: top;
    color: $secondary;
  }
}

.from,
.to {
  padding-top: $table-cell-padding-y * 0.5;
  padding-bottom: $table-cell-padding-y * 0.5;
}

.address {
  display: block;
}

.amount {
  color: $secondary;
}
