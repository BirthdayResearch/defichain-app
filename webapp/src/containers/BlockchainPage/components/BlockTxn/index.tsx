import React, { useState } from 'react';
import { Button, Card } from 'reactstrap';
import { MdInfo, MdArrowForward } from 'react-icons/md';
import styles from './BlockTxn.module.scss';
import CopyToClipboard from '../../../../components/CopyToClipboard';
import classnames from 'classnames';
import { I18n } from 'react-redux-i18n';
import { ITxn } from '../../interfaces';

interface BlockTxnProps {
  txn: ITxn;
  unit: string;
}

interface BlockTxnState {
  copied: boolean;
}

const BlockTxn: React.FunctionComponent<BlockTxnProps> = (
  props: BlockTxnProps
) => {
  const { txn, unit } = props;
  const [copied, setCopied] = useState(false);

  const handleCopy = () => {
    setCopied(true);
    setTimeout(() => {
      setCopied(false);
    }, 600);
  };
  return (
    <Card className={`${styles.txn} styles.txn`}>
      <div className={styles.header}>
        <span className={styles.hash}>
          <div
            className={classnames(
              { 'd-block': copied },
              styles.copiedIndicator
            )}
          >
            {I18n.t('containers.blockChainPage.blockTxn.copied')}
          </div>
          <span className={styles.hashSpan}>{txn.hash}</span>
          <CopyToClipboard value={txn.hash} handleCopy={handleCopy} />
        </span>
        <span className={styles.time}>
          <span className={styles.timeSpan}>{txn.time}</span>
          <Button color='link' size='sm' className='padless ml-2'>
            <MdInfo />
          </Button>
        </span>
      </div>
      <div className={styles.addresses}>
        <div className={styles.froms}></div>
        <div className={styles.arrow}>
          <MdArrowForward />
        </div>
        <div className={styles.tos}>
          {txn.tos.map((to, index) => (
            <div className={styles.to} key={`${to.address}${index}`}>
              <span className={styles.address}>
                {to.address
                  ? to.address
                  : I18n.t(
                      'containers.blockChainPage.blockTxn.unparsedAddress'
                    )}
              </span>
              <span className={styles.amount}>
                {to.amount}&nbsp;
                <span className={styles.unit}>{unit}</span>
              </span>
            </div>
          ))}
        </div>
      </div>
    </Card>
  );
};

export default BlockTxn;
