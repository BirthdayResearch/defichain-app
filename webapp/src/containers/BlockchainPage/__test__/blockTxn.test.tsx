import React from 'react';
import BlockTxn from '../components/BlockTxn';
import { shallow } from 'enzyme';
const txn = {
  hash: '2446f1fd773fbb9f080e674b60c6a033c7ed7427b8b9413cf28a2a4a6da9b56c',
  time: '1591108054',
  froms: [
    {
      address: 'bcrt1qw2grcyqu9jfdwgrggtpasq0vdtwvecty4vf4jk',
      amount: 10,
    },
  ],
  tos: [
    {
      address: 'bcrt1qw2grcyqu9jfdwgrggtpasq0vdtwvecty4vf28k',
      amount: 1,
    },
  ],
  unit: 'DFI',
};

describe('BlockTxn component', () => {
  it('should check for snapshot', () => {
    const wrapper = shallow(<BlockTxn txn={txn} unit={'DFI'} />);
    expect(wrapper).toMatchSnapshot();
  });
});
