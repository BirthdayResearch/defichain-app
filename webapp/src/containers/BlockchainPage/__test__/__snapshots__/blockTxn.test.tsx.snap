// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockTxn component should check for snapshot 1`] = `
<Card
  className="undefined styles.txn"
  tag="div"
>
  <div>
    <span>
      <div
        className=""
      >
        Copied!
      </div>
      <span>
        2446f1fd773fbb9f080e674b60c6a033c7ed7427b8b9413cf28a2a4a6da9b56c
      </span>
      <CopyToClipboard
        handleCopy={[Function]}
        value="2446f1fd773fbb9f080e674b60c6a033c7ed7427b8b9413cf28a2a4a6da9b56c"
      />
    </span>
    <span>
      <span>
        1591108054
      </span>
      <Button
        className="padless ml-2"
        color="link"
        size="sm"
        tag="button"
      >
        <MdInfo />
      </Button>
    </span>
  </div>
  <div>
    <div />
    <div>
      <MdArrowForward />
    </div>
    <div>
      <div
        key="bcrt1qw2grcyqu9jfdwgrggtpasq0vdtwvecty4vf28k0"
      >
        <span>
          bcrt1qw2grcyqu9jfdwgrggtpasq0vdtwvecty4vf28k
        </span>
        <span>
          1
          Â 
          <span>
            DFI
          </span>
        </span>
      </div>
    </div>
  </div>
</Card>
`;
